server/api/handler/transaction.go

```diff
@@ -0,0 +1,301 @@
+				return fmt.Errorf("failed to update webauthn credential: %w", err)
+			}
+		}
+
+		err = sessionDataPersister.Delete(*sessionData)
+		if err != nil {
+			ctx.Logger().Error(err)
+			return fmt.Errorf("failed to delete assertion session data: %w", err)
+		}
+
+		generator := ctx.Get("jwt_generator").(jwt.Generator)
+		token, err := generator.GenerateForTransaction(webauthnUser.UserId, base64.RawURLEncoding.EncodeToString(credential.ID), transaction.Identifier)
+		if err != nil {
+			ctx.Logger().Error(err)
+			return fmt.Errorf("failed to generate jwt: %w", err)
+		}
+
+		auditErr := h.AuditLog.CreateWithConnection(tx, models.AuditLogWebAuthnTransactionFinalSucceeded, &webauthnUser.UserId, transaction, nil)
+		if auditErr != nil {
+			ctx.Logger().Error(auditErr)
+			return fmt.Errorf(auditlog.CreationFailureFormat, auditErr)
+		}
+
+		return ctx.JSON(http.StatusOK, &response.TokenDto{Token: token})
+	})
+}
+
+func (t *transactionHandler) withTransaction(transactionId string, transactionDataJson string) webauthn.LoginOption {
+	return func(options *protocol.PublicKeyCredentialRequestOptions) {
+		transaction := []byte(transactionId)
+		transactionData := []byte(transactionDataJson)
+
+		transaction = append(transaction, transactionData[:]...)
+
+		transactionHash := sha256.Sum256(transaction)
+		options.Challenge = append(options.Challenge, transactionHash[:]...)
+	}
+}
+
+func (t *transactionHandler) getTransactionByChallenge(challenge string, persister persisters.TransactionPersister, tenantId uuid.UUID) (*models.Transaction, error) {
+	transaction, err := persister.GetByChallenge(challenge, tenantId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get transaction data: %w", err)
+	}
+
+	if transaction == nil {
+		return nil, echo.NewHTTPError(http.StatusNotFound, "no transaction found for this challenge")
+	}
+
+	return transaction, nil
+}
+
+func (t *transactionHandler) getSessionDataByChallenge(challenge string, persister persisters.WebauthnSessionDataPersister, tenantId uuid.UUID) (*models.WebauthnSessionData, error) {
+	sessionData, err := persister.GetByChallenge(challenge, tenantId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get webauthn assertion session data: %w", err)
+	}
+
+	if sessionData != nil && sessionData.Operation != models.WebauthnOperationAuthentication {
+		sessionData = nil
+	}
+
+	if sessionData == nil {
+		return nil, echo.NewHTTPError(http.StatusUnauthorized, "received challenge does not match with any stored one")
+	}
+
+	return sessionData, nil
+}
+
+func (t *transactionHandler) getWebauthnUserByUserHandle(userHandle string, tenantId uuid.UUID, persister persisters.WebauthnUserPersister) (*intern.WebauthnUser, error) {
+	user, err := persister.GetByUserId(userHandle, tenantId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get user: %w", err)
+	}
+
+	if user == nil {
+		return nil, fmt.Errorf("user not found")
+	}
+
+	return intern.NewWebauthnUser(*user), nil
+}
```

Comment:
This function can be moved to `webauthnHandler` because it is used and declared in transaction and login.

Reply:
Fixed